src/Server/Repositories/ClaraJobRepository.cs

```diff
@@ -0,0 +1,241 @@
+﻿/*
+ * Apache License, Version 2.0
+ * Copyright 2019-2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using Ardalis.GuardClauses;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Nvidia.Clara.DicomAdapter.API;
+using Nvidia.Clara.DicomAdapter.API.Rest;
+using Nvidia.Clara.DicomAdapter.Common;
+using Nvidia.Clara.DicomAdapter.Configuration;
+using Polly;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Nvidia.Clara.DicomAdapter.Server.Repositories
+{
+    public class ClaraJobRepository : IJobRepository
+    {
+        private const int ERROR_HANDLE_DISK_FULL = 0x27;
+        private const int ERROR_DISK_FULL = 0x70;
+        private const int MaxRetryLimit = 3;
+
+        private readonly IOptions<DicomAdapterConfiguration> _configuration;
+        private readonly ILogger<ClaraJobRepository> _logger;
+        private readonly IFileSystem _fileSystem;
+        private readonly IDicomAdapterRepository<InferenceJob> _inferenceJobRepository;
+
+        public ServiceStatus Status { get; set; } = ServiceStatus.Unknown;
+
+        public ClaraJobRepository(
+            ILogger<ClaraJobRepository> logger,
+            IOptions<DicomAdapterConfiguration> configuration,
+            IFileSystem fileSystem,
+            IDicomAdapterRepository<InferenceJob> inferenceJobRepository)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _inferenceJobRepository = inferenceJobRepository ?? throw new ArgumentNullException(nameof(inferenceJobRepository));
+        }
+
+        public async Task Add(Job job, string jobName, IList<InstanceStorageInfo> instances)
```

Comment:
this method needs comments explaining what it is intending to do. I have no idea if it is doing the right things or not, it's very confusing as-is.

Reply:
Added comments.